src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs

```diff
@@ -102,18 +100,14 @@ public override bool GetMemberAndWriteJson(object obj, ref WriteStack state, Utf
             bool success;
             T value = Get!(obj);
 
-            if (value == null)
+            if (IgnoreDefaultValuesOnWrite && (
+                default(T) is null ? value is null : EqualityComparer<T>.Default.Equals(default, value)))
```

Comment:
From my perspective, this is now adding complexity that is only worth it if there's a measurable perf benefit when actually using this functionality via the JSON APIs and not just in a microbenchmark on EqualityComparer.

Reply:
As @steveharter I don't think it will give any visible difference because the serializer suffers from other problems, but it's better than before. Therefore, I'm leaving decision taking to @layomia and the rest of the team, they maintain code mostly and maybe have some plans on further improvement.

Reply:
> but it's better than before

How do we know that?

Reply:
I mean not the current improvement, but all changes since 3.0.

Reply:
Found that `CanBeNull` can be made static because `TypeToConvert` is sealed and always equals to `typeof(T)`.

Reply:
The updated condition is also wrong:
- nothing will be written when `IgnoreDefaultValuesOnWrite` is true and the value is not `default`.
- `default` values will be written regardless of  the value of `IgnoreDefaultValuesOnWrite`.
- maybe a couple more things

Reply:
TLDR: After measuring the perf impact using the JSON APIs, I think the added complexity is worth it.

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19041.329 (2004/?/20H1)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-preview.7.20303.12
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.30301, CoreFX 5.0.20.30301), X64 RyuJIT
  Job-SPWLSQ : .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT
```

**Simplified condition (no fast path for reference types)**

|    Method |     Mean |    Error |   StdDev |   Median |      Min |      Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |---------:|---------:|---------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| Serialize | 571.0 ns | 13.63 ns | 15.15 ns | 567.5 ns | 555.2 ns | 603.9 ns | 0.0783 |     - |     - |     328 B |

**Complex condition (fast path for reference types)**

|    Method |     Mean |   Error |  StdDev |   Median |      Min |      Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |---------:|--------:|--------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| Serialize | 533.0 ns | 8.09 ns | 7.57 ns | 529.7 ns | 520.2 ns | 544.4 ns | 0.0770 |     - |     - |     328 B |

summary:
better: 1, geomean: 1.071
total diff: 1

No Slower results for the provided threshold = 0% and noise filter = 0.3ns.

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.IgnoreDefaultValues<LargeStructWithProperti |      1.07 |           567.52 |           529.69 |         |

```cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using BenchmarkDotNet.Attributes;
using MicroBenchmarks;
using MicroBenchmarks.Serializers;

namespace System.Text.Json.Serialization.Tests
{
    [GenericTypeArguments(typeof(LargeStructWithProperties))]
    public class IgnoreDefaultValues<T>
    {
        private T _value;
        private JsonSerializerOptions _options;

        [GlobalSetup]
        public void Setup()
        {
            _value = DataGenerator.Generate<T>();
            _options = new JsonSerializerOptions { DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault };
        }

        [BenchmarkCategory(Categories.Libraries, Categories.JSON)]
        [Benchmark]
        public byte[] Serialize() => JsonSerializer.SerializeToUtf8Bytes(_value, _options);
    }

    public struct LargeStructWithProperties
    {
        public string String1 { get; set; }
        public string String2 { get; set; }
        public string String3 { get; set; }
        public string String4 { get; set; }
        public string String5 { get; set; }
        public int Int1 { get; set; }
        public int Int2 { get; set; }
        public int Int3 { get; set; }
        public int Int4 { get; set; }
        public int Int5 { get; set; }
    }
}
```

Reply:
Thanks, @layomia.  In your repro, all of the strings are null.  What results do you get if they're not?

Reply:
Actually, that's the case where the strings are not null

```cs
private static LargeStructWithProperties CreateLargeStructWithProperties()
  => new LargeStructWithProperties
  {
      String1 = "1",
      String2 = "2",
      String3 = "3",
      String4 = "4",
      String5 = "5",
  };
```

---

Here are the numbers when they are null (which is what I meant to measure above :) -

**Simplified condition (no fast path for reference types)**

|    Method |     Mean |   Error |  StdDev |   Median |      Min |      Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |---------:|--------:|--------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| Serialize | 303.8 ns | 4.52 ns | 4.22 ns | 302.7 ns | 298.2 ns | 312.4 ns | 0.0625 |     - |     - |     264 B |

**Complex condition (fast path for reference types)**

|    Method |     Mean |   Error |  StdDev |   Median |      Min |      Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |---------:|--------:|--------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| Serialize | 253.8 ns | 3.66 ns | 3.42 ns | 253.0 ns | 248.5 ns | 260.0 ns | 0.0620 |     - |     - |     264 B |

summary:
better: 1, geomean: 1.196
total diff: 1

No Slower results for the provided threshold = 0% and noise filter = 0.3ns.

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.IgnoreDefaultValues<LargeStructWithProperti |      1.20 |           302.67 |           253.00 |

```cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using BenchmarkDotNet.Attributes;
using MicroBenchmarks;
using MicroBenchmarks.Serializers;

namespace System.Text.Json.Serialization.Tests
{
    [GenericTypeArguments(typeof(LargeStructWithProperties))]
    public class IgnoreDefaultValues<T>
    {
        private T _value;
        private JsonSerializerOptions _options;

        [GlobalSetup]
        public void Setup()
        {
            _value = default;
            _options = new JsonSerializerOptions { DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault };
        }

        [BenchmarkCategory(Categories.Libraries, Categories.JSON)]
        [Benchmark]
        public byte[] Serialize() => JsonSerializer.SerializeToUtf8Bytes(_value, _options);
    }
}
```